### ISSUE TITLE
beta v1.7.0 python packaged version

### ISSUE THREAD

**Issue by nextgenusfs** [nextgenusfs/funannotate#343]
Hi @hyphaltip @PlantDr430.  I think I've gotten nearly 99% of the way to re-packaging code so that it can be installed via conda (ie is a proper python package).  This is only python2 at the moment.  So @PlantDr430 once we get this working, then can start on the py2/3 migration. 

These data are on the development branch.  I've also implemented a re-write for `funannotate species` and `funannotate predict` that allows for re-using the ab-initio training data from all of the prediction software.  So this starts with `funannotate setup` which will copy over all Augustus pre-trained datasets into $FUNANNOTATE_DB/trained_species directory.  The data (locations) of training parameters are stored here and `funannotate predict` parses the JSON file in each directory to get the appropriate files to run each program.  This means that we are doing all training/predicting in the local directory and not $AUGUSTUS_CONFIG_DIR or $FUNANNOTATE_DB.  The parameter JSON file can be passed to `funannotate predict -p species.json` in order for a user to re-use a parameters set.  They can then be added to the database by using the `funannotate species` script -- and then be used like previously.

It would be great if you could test the development version.  It can be installed with pip.  The goal here is to make a conda package so we can install with `conda install funannotate`, so we should be able to do the following (please update if this doesn't work):
```
conda create -y -n funannotate "python=2.7" numpy pandas scipy matplotlib seaborn \
    natsort scikit-learn psutil biopython requests blast goatools fisher tantan \
    bamtools augustus bedtools hmmer exonerate "diamond>=0.9" tbl2asn ucsc-pslcdnafilter \
    samtools raxml trimal "mafft>=7" iqtree kallisto bowtie2 infernal minimap2  \
    "trinity>=2.6.6" evidencemodeler "pasa>=2.3" codingquarry stringtie "gmap=2017.11.15" snap \
    ete3 "salmon>=0.9" "jellyfish>=2.2" htslib trnascan-se hisat2 glimmerhmm \
    perl-threaded perl-db-file perl-bioperl perl-dbd-mysql perl-dbd-sqlite \
    perl-text-soundex perl-scalar-util-numeric perl-data-dumper perl-dbi perl-clone \
    perl-json perl-logger-simple perl-hash-merge perl-yaml perl-pod-usage perl-getopt-long \
    perl-parallel-forkmanager perl-carp perl-soap-lite perl-class-inspector perl-app-cpanminus \
    "proteinortho>=6.0.9"
```
And then install funannotate development version:
```
git checkout development
pip install .
```
Let me know if the above works/doesn't work/etc. 

Thanks,
Jon

UPDATE: removed -b from the git checkout command and double quoted the version.  Added proteinortho (proteinortho6)

UPDATE2: added `tantan` and removed repeatmodeler/masker required dependencies. We need a RepBase independent masking solutions.  Also removed `blat` -- as academic license only so won't be included. removed 'mummer' as clean now done with minimap2

**Comment by hyphaltip** [comment]
The ">=" copy and paste is triggering some weird redirection - modifying above with "trinity>=2.6.6"  eg seems to solve that so yo might want to update the text so it can be copy+pasteable.

**Comment by hyphaltip** [comment]
one small thing - i think the -b is not needed since remote branch already exists.  
```
$ git checkout -b development
fatal: A branch named 'development' already exists.
$ git checkout development
Switched to branch 'development'
$ ls -l 
ls -l
total 512
-rw-r--r-- 1 pkgadmin pkgadmin  1322 Oct 24 20:44 LICENSE.md
-rw-r--r-- 1 pkgadmin pkgadmin   279 Oct 24 21:26 MANIFEST.in
-rw-r--r-- 1 pkgadmin pkgadmin   500 Oct 24 20:44 README.md
drwxr-xr-x 2 pkgadmin pkgadmin  4096 Oct 24 21:19 __pycache__
drwxr-xr-x 2 pkgadmin pkgadmin  4096 Oct 24 20:44 dockerbuild
drwxr-xr-x 2 pkgadmin pkgadmin  4096 Oct 24 20:44 docs
drwxr-xr-x 6 pkgadmin pkgadmin  4096 Oct 24 21:26 funannotate
-rw-r--r-- 1 pkgadmin pkgadmin 37960 Oct 24 21:25 funannotate.pyc
drwxr-xr-x 2 pkgadmin pkgadmin  4096 Oct 24 21:26 lib
drwxr-xr-x 2 pkgadmin pkgadmin  4096 Oct 24 20:44 sample_data
drwxr-xr-x 2 pkgadmin pkgadmin  4096 Oct 24 21:26 scripts
-rwxr-xr-x 1 pkgadmin pkgadmin  3818 Oct 24 21:26 setup.py
$ pip install .
pip install .
DEPRECATION: Python 2.7 will reach the end of its life on January 1st, 2020. Please upgrade your Python as Python 2.7 won't be maintained after that date. A future version of pip will drop support for Python 2.7. More details about Python 2 support in pip, can be found at https://pip.pypa.io/en/latest/development/release-process/#python-2-support
...
Successfully installed atomicwrites-1.3.0 attrs-19.3.0 configparser-4.0.2 contextlib2-0.6.0.post1 coverage-4.5.4 docopt-0.6.2 funannotate-1.7.0 funcsigs-1.0.2 importlib-metadata-0.23 more-itertools-5.0.0 nose-1.3.7 packaging-19.2 pathlib2-2.3.5 pluggy-0.13.0 py-1.8.0 pytest-4.6.6 pytest-cov-2.8.1 scandir-1.10.0 wcwidth-0.1.7 xlrd-1.2.0 zipp-0.6.0
$ which funannotate
/opt/linux/centos/7.x/x86_64/pkgs/anaconda3/4.5.4/envs/funannotate/bin/funannotate
```

Seems to work for installation will try out some runs - we generally have a global shared $FUNANNOTATE_DB on our cluster so I need to see what makes sense for the shared access approach (make local project folder copy of FUNANNOTATE_DB and symlink back to the main DBs for example)

**Comment by nextgenusfs** [comment]
The training files are only saved locally in the run each time. I was thinking that would work in your system as well, but to add a new species to be available to every user would need to be added by a permissioned user and the funannotate species command.  Certainly open to a different method if you have a better idea.

**Comment by nextgenusfs** [comment]
I also know that the `CodingQuarry` conda recipe is setting `$QUARRY_PATH` to the wrong location -- so if CQ is falling that is why.  I need to fix the conda recipe for CQ.

**Comment by PlantDr430** [comment]
I am glad this is moving forward. I haven't had yet to try the install
myself, but I will soon. Once I do that I will try to do a run through
myself to look for any errors.

**Comment by hyphaltip** [comment]
I guess on our shared system I could just leave off many of the conda pkg installs for primary apps (augustus, codingquarry, minimap2 etc) if we have them loaded and installed already. 

I guess maybe there could be a thought that either 
a) shared install of immutable dbs (uniprot, pfam, merops,cazy) in a global directory
b) global install of augustus set parameters transferred to the JSON packaged folder also in immutable folder
c) local install could be supported ~/.funannotate/DB not unlike how conda supports global and local envs builds. 

If global exists, (set by $FUNANNOTATE_DB) use it and search for dbs/augustus params. If global is NOT writeable then try to write new params there?

For a docker install or user who has installed everything in their 

The issues are when you are trying to satisfy a dependency you just need to have a search order to find the file you want - decide on the precendence. If people have outdated global uniprot db - can the override it with a local copy, etc?

**Comment by nextgenusfs** [comment]
So I think I tried to sort of do what you are describing... maybe?  On the run of `funannotate setup` the script parses the existing species in `$AUGUSTUS_CONFIG_PATH/species` and moves parameters to `$FUNANNOTATE_DB` and writes a JSON file describing what is in there to `$FUNANNOTATE_DB/trained_species/species_name/info.json`.

When `funannotate predict` is run it then searches if species is available in `$FUNANNOTATE_DB/trained_species`, if so, then parses the `info.json` file to copy those training parameters to the local run folder.  It then proceeds with the "best" approach, ie if pre-training data exists for all ab-initio predictors, it will just use those parameters.  If it needs to train any of those predictors it will do that logically (if PASA exists will use those data, otherwise will run BUSCO). When `funannotate predict` finishes it will write the local JSON parameters to `predict_results/species.parameters.json`.  It does not try to push this back to $FUNANNOTATE_DB.

The commmand `funannotate species -s species_name -a species.parameters.json` can be run to add a training set to the database.  This script will check for write access to that location before it does anything, if no write access will just exit with a 'no permissions' error.  

The local parameters JSON file can also be used directly by `funannotate predict`, ie you can run it like:
```
funannotate predict -p species.parameters.json -i genome.fasta -s "Species greaticus"
```
If the `-p, --parameters` option is passed it will over-ride any fetching of the species name from $FUNANNOTATE_DB. 

Hopefully this works with users that may not have write access to a shared/immutable $FUNANNOTATE_DB?

**Comment by hyphaltip** [comment]
yes - I think this captures the idea. Just trying to think about how you could pre-load a system with all of what is need (eg docker containers/img we could deploy and have all of what is needed). but then also have ability to pull in customized parameter builds since presumably that will be a common need too.

**Comment by hyphaltip** [comment]
just that there is no need for every user to have complete copy of the db folder in some sort of writeable local folder since that is just wasted space and time to download.

**Comment by nextgenusfs** [comment]
Okay, yeah its not doing that currently -- just the training parameters.

**Comment by nextgenusfs** [comment]
Also -- I was not planning on continuing to build the docker container if I'm able to get a conda recipe working.  Or I guess we could try to get continuous integration working so that it would be created automatically -- I just don't really care for docker and in this case ends up with a quite large image that end users have to build themselves anyway.

**Comment by hyphaltip** [comment]
Yes - I guess I was speaking more about the general concept of a clean image that could be deployed on an HPCC / Cyverse etc.

**Comment by hyphaltip** [comment]
I ran into this problem for CodingQuarry bioconda install - the config files from the page are going into
 ... anaconda3/4.5.4/envs/funannotate/opt/codingquarry-2.0/QuarryFiles/QuarryFiles/
but should be 
 ... anaconda3/4.5.4/envs/funannotate/opt/codingquarry-2.0/QuarryFiles/

we should issue a bioconda PR?

**Comment by nextgenusfs** [comment]
I think I fixed that a few days ago — try to upgrade that package.

**Comment by hyphaltip** [comment]
We are going to try to build singularity container for my runs here on HPCC and if that works might be simpler for certain deploying - I have goals of trying to make this work on atmosphere/Cyverse.

**Comment by nextgenusfs** [comment]
Okay so that is supposedly better than docker, correct?

**Comment by hyphaltip** [comment]
Our Hpcc admin says so. I will give it a try. I’m not yet sold of it for our workflow but let’s see.

Jason Stajich, PhD
jasonstajich.phd@gmail.com

**Comment by nextgenusfs** [comment]
Conda package now released, let me know if works/doesn't work for you.  Next step is py2/3 conversion.
